crates/chain-state/src/notifications.rs

```diff
@@ -211,6 +212,86 @@ impl<T: Clone + Sync + Send + 'static> Stream for ForkChoiceStream<T> {
     }
 }
 
+/// A stream that emits canonical chain notifications only when blocks are finalized.
+#[pin_project]
+pub(crate) struct FinalizedNotificationStream<N: NodePrimitives> {
+    #[pin]
+    canon_stream: CanonStateNotificationStream<N>,
+    #[pin]
+    finalized_stream: ForkChoiceStream<SealedHeader<alloy_consensus::Header>>,
+    // Buffer for the most recent canonical notification
+    buffered_notification: Option<CanonStateNotification<N>>,
+    // Buffer for the most recent finalization event
+    buffered_finalization: Option<SealedHeader<alloy_consensus::Header>>,
+}
+
+impl<N: NodePrimitives> FinalizedNotificationStream<N>
+where
+    N::Block: Clone + Send + Sync + 'static,
+{
+    // Helper method to check if we have a matching pair of notifications
+    fn try_match(
+        buffered_notification: &Option<CanonStateNotification<N>>,
+        buffered_finalization: &Option<SealedHeader<alloy_consensus::Header>>,
+    ) -> Option<CanonStateNotification<N>> {
+        match (buffered_notification, buffered_finalization) {
+            (Some(notification), Some(finalized_header))
+                if notification.tip().hash() == finalized_header.hash() =>
+            {
+                Some(notification.clone())
+            }
+            _ => None,
+        }
```

Comment:
I'd like to inline this function so that we can .take the notification instead.

so we can move this inside the poll fn